### ISSUE TITLE
cmd/go: incorrectly rejects '@' character in directory names

### ISSUE THREAD

**Issue by jayconrod** [golang/go#35115]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go1.13.3
</pre>

### Does this issue reproduce with the latest release?

Reproduces in go1.13.3 but not in go1.13.2 or go1.12.11. This seems to be a regression.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/jayconrod/Library/Caches/go-build"
GOENV="/Users/jayconrod/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/jayconrod/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/opt/go/installed"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/opt/go/installed/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/jayconrod/Code/test/@/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/rq/x0692kqj6ml8cvrhcqh5bswc008xj1/T/go-build027955611=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

```
mkdir @
cd @
go mod init m
echo 'package m' >m.go
go list $(pwd)
```

### What did you expect to see?

Same output as `go list .`:

```
m
```

### What did you see instead?

```
can't load package: package /Users/jayconrod/Code/test@: can only use path@version syntax with 'go get'
```

The `@` character is not valid in module paths or package paths. However, it is allowed in file names. It should certainly be allowed in module parent directory names.

**Comment by jayconrod** [comment]
Caused by [CL 194600](https://go-review.googlesource.com/c/go/+/194600)

**Comment by jayconrod** [comment]
@gopherbot Please backport to go1.13

**Comment by zrhmn** [comment]
For other people following this issue, the problem persists even with go1.13.4.

The fix was not backported to 1.13 because [the cherrypick was rejected](https://github.com/golang/go/issues/35119). I would have preferred to be a part of that discussion, but it's my own fault for not following the maze of issue-links. Too late now. 🤷‍♂

The fix is [here](https://go-review.googlesource.com/c/go/+/202763/1/src/cmd/go/internal/search/search.go#368) for anyone interested. May still make it in go1.14.
